From 98e33cc11ed6829c3261d50252f550e40dcc0613 Mon Sep 17 00:00:00 2001
From: Neti Ravi Kumar <ravineti@codeaurora.org>
Date: Wed, 13 Feb 2013 18:12:42 +0530
Subject: [PATCH] gralloc : Don't fallback to system heap for framebuffer
 allocations

MDP can not handle buffers from system heap, so try camera heap
first, before falling back completely to system heap

Change-Id: I7e2ab0094be5d1075be1ca93295cf54b84101835
---
 libgralloc/alloc_controller.cpp | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/libgralloc/alloc_controller.cpp b/libgralloc/alloc_controller.cpp
index 46f6f9b..57f28b8 100644
--- a/libgralloc/alloc_controller.cpp
+++ b/libgralloc/alloc_controller.cpp
@@ -265,7 +265,19 @@ int IonController::allocate(alloc_data& data, int usage)
     data.flags = ionFlags;
     ret = mIonAlloc->alloc_buffer(data);
 
-    // Fallback
+    // Fallback to CAMERA PREVIEW heap if
+    // allocation is for fb1 AND
+    // allocation from SF heap fails
+    // Reason : fb1 had to go to MDP and MDP can't use system heap
+    // TODO : don't hardcode based on 'usage'
+
+    if(ret < 0 && (usage == 6659)) {
+       ALOGW("Falling back to CAMERA_PREVIEW heap for fb allocations");
+       data.flags = ION_HEAP(ION_CAMERA_HEAP_ID);
+       ret = mIonAlloc->alloc_buffer(data);
+    }
+
+    // Fallback to system heap
     if(ret < 0 && canFallback(usage,
                               (ionFlags & ION_SYSTEM_HEAP_ID)))
     {
